### DISCUSSION TITLE
support for deno

### DISCUSSION THREAD

**Discussion by ivanhofer** [ivanhofer/typesafe-i18n#87] [Ideas]
[deno](https://deno.land/) is an alternative to Node.js. It has TypeScript support out-of-the-box. It would be nice to be able to use `typesafe-i18n` with deno.
 - [x] investigate what is needed to add support for deno
    - all imports need to include the file-ending `*.ts`:
      - [x] update file generation to include file endings when outputting for deno (`adapter: 'deno'`)
    - how can we import (only) types in deno??
       - [x] **find a solution to that problem**
 - [ ] re-write watcher to be able to run with deno instead of node

**Comment by zabihexes** [comment]
btw it works with deno now through deno npm specifiers.

**Reply by ivanhofer** [reply]
Thanks for checking!

**Comment by aW4KeNiNG** [comment]
Yes, I can confirm that it works on Deno. But the problem is that the Deno resolver can't find the module because the extension is missing in the imports. It compiles, but the intellisense is missing with this error:

```Deno: Unable to load a local module: "./src/i18n/i18n-util". Please check the file path.```

If I add the `.ts` extension in all imports it works fine and I can see intellisense for VSCode or IntelliJ.

For example, the `i18n-util.sync.ts`file:

```
import { initFormatters } from './formatters.ts';
import type { Locales, Translations } from './i18n-types.ts';
import { loadedFormatters, loadedLocales, locales } from './i18n-util';

import de from './de/index.ts';
import en from './en/index.ts';
import it from './it/index.ts';
```

But, ofc, I can't overwrite them because these files are autogenerated.

**Reply by ivanhofer** [reply]
did you try to set the [`esmImports`-option](https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/generator#esmimports) to `true`?

**Reply by aW4KeNiNG** [reply]
Oh, I missed that option! I have tried it, but it is adding the `.js` extension and Deno can't resolve it (all files have the `.ts` extension). 

And I have another error if I change the import statement in the `en/index.ts` file: `import type { BaseTranslation } from '../i18n-types.ts';`
I get the next error when I run: `deno run -A npm:typesafe-i18n`

```
deno run -A --unstable --node-modules-dir npm:typesafe-i18n
[typesafe-i18n] version 5.21.0
[typesafe-i18n] generating files for TypeScript version: '4.9.x'
[typesafe-i18n] options: { adapter: "solid", esmImports: true, outputPath: "./src-ui/i18n/" }
[typesafe-i18n] watcher started in: './src-ui/i18n/'
[typesafe-i18n] ERROR: in 'en'
Make sure to import the type 'BaseTranslation' from the generated 'i18n-types.ts' file.
See the example in the official docs: https://github.com/ivanhofer/typesafe-i18n/tree/main/packages/generator#namespaces

[typesafe-i18n] generated file: './src-ui/i18n/i18n-types.ts'
[typesafe-i18n] ... all files are up to date
```

This error doesn't happen if the import is `../i18n-types.js` or `../i18-types`. Weird :S

Also, the changes in the translation files doesn't trigger the watcher to update the autogenerated files.

Thanks in advance!

**Reply by ivanhofer** [reply]
Is this the default way how `deno` handles `TypeScript` imports?
In a regular `type="module"` `TypeScript` project you need to use the `.js` file extension even if you are importing something from a `.ts` file.

**Reply by aW4KeNiNG** [reply]
Yes, you cannot omit the file extension and Deno does not perform any "magic" resolution. If you specify .js, it will attempt to locate a .js file. [Here](https://github.com/denoland/deno/issues/2506) is a discussion about the reasons for doing it explicitly.

**Reply by ivanhofer** [reply]
It totally makes sense so specify the real file extension and I find the node behavior weird ^^.
So this means a new option is needed to output `deno` code. Is there anything else that is needed in order to make `typesafe-i18n` work?
Would you like to create a small repository? Maybe with the first commit where you install `typesafe-i18n` regulary and then a second commit with the changes necessary to make it work? Then we can take a look at the diff to see what changes are needed.

**Reply by aW4KeNiNG** [reply]
Ok. I'll give it a try tomorrow and post the result here :)

**Reply by aW4KeNiNG** [reply]
Here the repository: https://github.com/aW4KeNiNG/deno-solidjs-vite-i18n-template

Another issue. When I run `deno run -A npm:typesafe-i18n --setup`, the `.typesafe-i18n.json` is created but it throws the next error: 
[typesafe-i18n] generated config file: '.typesafe-i18n.json'
[typesafe-i18n] ERROR: no 'package.json' found. You have to install 'typesafe-i18n' by yourself

Maybe it could detect the deno.json file (or deno.jsonc) too.

**Reply by ivanhofer** [reply]
Will take a look at this when I have time

**Reply by ivanhofer** [reply]
@aW4KeNiNG I have hacked together something and released version `5.24.0`. Please check if your points are resolved.
As I have never used `deno` before, I'm not really sure how to update to the latest version and to test it ^^

**Reply by aW4KeNiNG** [reply]
Thanks!! It works :)

The only issue I'm encountering is that changing a string or adding a new one in the location file does not regenerate the files as it does in Node. It only works the first time the command is executed. Do you have any idea what might be causing this problem?

Note: you can update deno using `deno upgrade`.

**Reply by ivanhofer** [reply]
> The only issue I'm encountering is that changing a string or adding a new one in the location file does not regenerate the files as it does in Node. It only works the first time the command is executed. Do you have any idea what might be causing this problem?

The generator uses `chokidar`. Maybe it does not detect changes when running in a `deno` process. You will probably need to prefix your command with a watcher package that is compatible with deno and disable the built-in watcher via [`--no-watch`](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/README.md#running-the-generator-in-cicd).

As I'm not aware of the `deno` ecosystem, I have left out some implementation details. It would be vreat if anyone could fill out any of the missing pieces:
 - get a list of all installed dependencies (see `TODO`s [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/cli/src/setup/runtimes/deno.mts)) see the `node` implementation as a reference
 - add support for `deno.jsonc` (see [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/cli/src/setup/runtimes/deno.mts#L13))
 - write docs similar to those for the `node`-adapter [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/adapter-node/README.md)
 - create a small example repository

**Reply by aW4KeNiNG** [reply]
Ok, thanks. It is weird, because sometimes it seems to work until I kill the process. I will check if I see something else or find the issue.

**Reply by zabihexes** [reply]
> > The only issue I'm encountering is that changing a string or adding a new one in the location file does not regenerate the files as it does in Node. It only works the first time the command is executed. Do you have any idea what might be causing this problem?
> 
> The generator uses `chokidar`. Maybe it does not detect changes when running in a `deno` process. You will probably need to prefix your command with a watcher package that is compatible with deno and disable the built-in watcher via [`--no-watch`](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/README.md#running-the-generator-in-cicd).
> 
> As I'm not aware of the `deno` ecosystem, I have left out some implementation details. It would be vreat if anyone could fill out any of the missing pieces:
> 
>     * get a list of all installed dependencies (see `TODO`s [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/cli/src/setup/runtimes/deno.mts)) see the `node` implementation as a reference
> 
>     * add support for `deno.jsonc` (see [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/cli/src/setup/runtimes/deno.mts#L13))
> 
>     * write docs similar to those for the `node`-adapter [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/adapter-node/README.md)
> 
>     * create a small example repository

I do not think detection Deno by looking for deno.json is a good idea. It can be `.json` or `.jsonc` btw. Also it is not required. You can use deno without it. You can even name the deno file differently....
I would suggest doing something like this instead. 
https://github.com/aaronccasanova/aacc/blob/main/packages/cross-runtime/src/utils.js#L33

**Reply by ivanhofer** [reply]
> I would suggest doing something like this instead.
> [aaronccasanova/aacc@`main`/packages/cross-runtime/src/utils.js#L33](https://github.com/aaronccasanova/aacc/blob/main/packages/cross-runtime/src/utils.js#L33)

Perfect, thanks for the hint! That is exactly what I was looking for.
